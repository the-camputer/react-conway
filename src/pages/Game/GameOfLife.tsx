import React, { useEffect, useState } from 'react';
import {
  Stack,
  Typography,
  Box,
  Button,
  Slider,
  FormControl,
  FormLabel,
} from '@mui/joy';
import { Clear, Replay, PlayArrow, Pause } from '@mui/icons-material';
import GoLField from './GoLField';
import { Cell, calculateNextState } from './GameOfLifeService';

const env = process.env.NODE_ENV;

const defaultCellSize = 40;

const GameOfLife: React.FC = (props) => {
  const [tick, setTick] = useState<number>(0);
  const [paused, setPaused] = useState<boolean>(true);
  const [cellSize, setCellSize] = useState<number>(defaultCellSize);
  const [gameState, setGameState] = useState<Cell[]>([
    { x: 15, y: 2 },
    { x: 16, y: 3 },
    { x: 14, y: 4 },
    { x: 15, y: 4 },
    { x: 16, y: 4 },
  ]);
  const [startingGameState, setStartingGameState] = useState<Cell[]>(gameState);

  useEffect(() => {
    if (tick === 0) {
      setStartingGameState(gameState);
    }
  }, [gameState, tick]);

  useEffect(() => {
    const interval = setInterval(() => {
      if (!paused) {
        setTick((tick) => tick + 1);
      }
    }, 500);
    return () => clearInterval(interval);
  }, [setTick, paused]);

  useEffect(() => {
    if (tick > 0) {
      setGameState((gameState) => calculateNextState(gameState));
    }
  }, [tick]);

  const toggleCell = (clicked: Cell) => {
    const clickedIndex = gameState.findIndex(
      (cell) => cell.x === clicked.x && cell.y === clicked.y
    );
    if (clickedIndex >= 0) {
      setGameState(gameState.filter((_cell, index) => index !== clickedIndex));
    } else {
      setGameState([...gameState, { ...clicked }]);
    }
  };

  const reset = () => {
    setGameState(startingGameState);
    setTick(0);
  };

  const clear = () => {
    setGameState([]);
  };

  const scaleCell = (
    _event: Event,
    value: number | number[],
    _activeThumb: number
  ) => {
    if (Array.isArray(value)) {
      setCellSize(value[0]);
    } else {
      setCellSize(value);
    }
  };

  return (
    <Stack spacing={0} justifyContent='center'>
      <Typography level='display1' variant='solid' id='title'>
        Conway's Game of Life
      </Typography>
      <GoLField
        cellSize={cellSize}
        livingCells={gameState}
        toggleFn={toggleCell}
      />
      {env === 'test' && (
        <div data-testid='game-data'>{JSON.stringify(gameState)}</div>
      )}
      <Box
        sx={{
          display: 'flex',
          gap: 2,
          margin: '10px 0 10px 0',
          flexWrap: 'wrap',
          justifyContent: 'center',
        }}
      >
        {!paused && (
          <Button
            size='lg'
            startDecorator={<Pause />}
            onClick={() => setPaused(true)}
            data-testid='pause-game'
          >
            Pause
          </Button>
        )}
        {paused && (
          <Button
            size='lg'
            startDecorator={<PlayArrow />}
            onClick={() => setPaused(false)}
            data-testid='play-game'
          >
            Play
          </Button>
        )}
        {tick === 0 && (
          <Button
            size='lg'
            startDecorator={<Clear />}
            onClick={clear}
            data-testid='clear-game'
          >
            Clear
          </Button>
        )}
        {tick > 0 && (
          <Button
            size='lg'
            startDecorator={<Replay />}
            onClick={reset}
            data-testid='reset-game'
          >
            Reset
          </Button>
        )}
      </Box>
      <Box
        sx={{
          display: 'flex',
          gap: 10,
          margin: '10px 10px 10px 10px',
          flexWrap: 'wrap',
          justifyContent: 'center',
        }}
      >
        <FormControl sx={{ flexGrow: 0 }}>
          <FormLabel>Cell Size</FormLabel>
          <Slider
            size='lg'
            defaultValue={defaultCellSize}
            min={10}
            max={100}
            step={2}
            onChange={scaleCell}
            sx={{ width: 200 }}
          />
        </FormControl>
      </Box>
    </Stack>
  );
};

export default GameOfLife;
