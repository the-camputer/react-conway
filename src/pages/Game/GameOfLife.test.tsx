import userEvent from '@testing-library/user-event';
import GameOfLife from './GameOfLife';
import { render, act, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

global.innerWidth = 1920;
global.innerHeight = 640;

describe('GameOfLife', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('renders game without crashing', () => {
    let view = render(<GameOfLife />, { wrapper: BrowserRouter });
    expect(view).toMatchSnapshot();
  });

  it('calcualtes a new game state on a regular cadence during play', async () => {
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
    render(<GameOfLife />, { wrapper: BrowserRouter });

    let gameState1 = JSON.parse(screen.getByTestId('game-data').textContent!);

    await act(async () => {
      await user.click(screen.getByTestId('play-game'));
    });

    act(() => {
      jest.advanceTimersByTime(500);
    });

    let gameState2 = JSON.parse(screen.getByTestId('game-data').textContent!);

    expect(gameState1[0]).toEqual(expect.objectContaining({ x: 15, y: 2 }));
    expect(gameState2[0]).toEqual(expect.objectContaining({ x: 14, y: 3 }));
  });

  it('adds a clicked cell that is currently dead to the game state to be rendered alive', async () => {
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
    render(<GameOfLife />, { wrapper: BrowserRouter });

    await act(async () => {
      await user.pointer({
        keys: '[MouseLeft]',
        coords: { clientX: 120, clientY: 160 },
        target: screen.getByTestId('gol-canvas'),
      });
    });

    let gameState = JSON.parse(screen.getByTestId('game-data').textContent!);
    expect(gameState).toEqual(
      expect.arrayContaining([expect.objectContaining({ x: 3, y: 4 })])
    );
  });

  describe('buttons', () => {
    describe('Pause & Play', () => {
      it('A play button appears when the game first starts', () => {
        render(<GameOfLife />, { wrapper: BrowserRouter });
        expect(screen.getByTestId('play-game')).toBeInTheDocument();
      });

      it('Play button switches to pause when pressed', async () => {
        const user = userEvent.setup({
          advanceTimers: jest.advanceTimersByTime,
        });
        render(<GameOfLife />, { wrapper: BrowserRouter });
        await act(async () => {
          await user.click(screen.getByTestId('play-game'));
        });

        expect(screen.getByTestId('pause-game')).toBeInTheDocument();
      });

      it('Pause button switches to play when pressed', async () => {
        const user = userEvent.setup({
          advanceTimers: jest.advanceTimersByTime,
        });

        render(<GameOfLife />, { wrapper: BrowserRouter });
        await act(async () => {
          await user.click(screen.getByTestId('play-game'));
        });

        await act(async () => {
          await user.click(screen.getByTestId('pause-game'));
        });

        expect(screen.getByTestId('play-game')).toBeInTheDocument();
      });
    });

    describe('Reset & Clear', () => {
      it('Clear button appears at tick 0 and reset button does not', async () => {
        render(<GameOfLife />, { wrapper: BrowserRouter });
        expect(screen.getByTestId('clear-game')).toBeInTheDocument();
        expect(screen.queryByTestId('reset-game')).not.toBeInTheDocument();
      });

      it('Clear button empties the game state', async () => {
        const user = userEvent.setup({
          advanceTimers: jest.advanceTimersByTime,
        });
        render(<GameOfLife />, { wrapper: BrowserRouter });
        const gameState1 = JSON.parse(
          screen.getByTestId('game-data').textContent!
        );
        await act(async () => {
          user.click(screen.getByTestId('clear-game'));
        });
        const gameState2 = JSON.parse(
          screen.getByTestId('game-data').textContent!
        );

        expect(gameState1.length).toBeGreaterThan(0);
        expect(gameState2.length).toBe(0);
      });

      it('Reset button sets the game state back to what it was at tick 0', async () => {
        const user = userEvent.setup({
          advanceTimers: jest.advanceTimersByTime,
        });

        render(<GameOfLife />, { wrapper: BrowserRouter });

        const gameState1 = JSON.parse(
          screen.getByTestId('game-data').textContent!
        );

        await act(async () => {
          await user.click(screen.getByTestId('play-game'));
        });

        act(() => {
          jest.advanceTimersByTime(501);
        });

        const gameState2 = JSON.parse(
          screen.getByTestId('game-data').textContent!
        );

        await act(async () => {
          await user.click(screen.getByTestId('reset-game'));
        });

        const gameState3 = JSON.parse(
          screen.getByTestId('game-data').textContent!
        );

        expect(gameState2).not.toEqual(gameState1);
        expect(gameState3).toEqual(gameState1);
      });
    });

    describe('Grid Manipulators', () => {
      describe('Cell Size', () => {
        it('starts with default size', () => {
          render(<GameOfLife />, { wrapper: BrowserRouter });
          const cellSize = parseInt(
            screen.getByTestId('cell-size').textContent!
          );
          expect(cellSize).toBe(40);
        });

        it('changes the size of the grid when the slider adjusts', async () => {
          const user = userEvent.setup({
            advanceTimers: jest.advanceTimersByTime,
          });

          render(<GameOfLife />, { wrapper: BrowserRouter });

          const sliderRail = screen.getByTestId('cell-size-rail');

          const originalCellSize = parseInt(
            screen.getByTestId('cell-size').textContent!
          );

          await act(async () => {
            await user.pointer({
              keys: '[MouseLeft]',
              target: sliderRail,
              coords: {
                x: sliderRail.clientLeft + 20,
                y: sliderRail.clientTop,
              },
            });
          });

          const newCellSize = parseInt(
            screen.getByTestId('cell-size').textContent!
          );

          expect(newCellSize).toBeGreaterThan(originalCellSize);
        });
      });

      describe('Update Speed', () => {
        it('starts with default speed', () => {
          render(<GameOfLife />, { wrapper: BrowserRouter });
          const cellSize = parseInt(
            screen.getByTestId('update-speed').textContent!
          );
          expect(cellSize).toBe(500);
        });

        it('changes the update speed when the slider adjusts', async () => {
          const user = userEvent.setup({
            advanceTimers: jest.advanceTimersByTime,
          });

          render(<GameOfLife />, { wrapper: BrowserRouter });

          const sliderRail = screen.getByTestId('update-speed-rail');

          const originalCellSize = parseInt(
            screen.getByTestId('update-speed').textContent!
          );

          await act(async () => {
            await user.pointer({
              keys: '[MouseLeft]',
              target: sliderRail,
              coords: {
                x: sliderRail.clientLeft + 20,
                y: sliderRail.clientTop,
              },
            });
          });

          const newCellSize = parseInt(
            screen.getByTestId('update-speed').textContent!
          );

          expect(newCellSize).toBeGreaterThan(originalCellSize);
        });
      });
    });
    describe('Import & Export', () => {
      describe('import', () => {
        it('button opens a modal to select file when clicked', async () => {
          const user = userEvent.setup({
            advanceTimers: jest.advanceTimersByTime,
          });
          render(<GameOfLife />, { wrapper: BrowserRouter });

          await act(async () => {
            await user.click(screen.getByTestId('open-import'));
          });

          expect(screen.getByTestId('import-select')).toBeInTheDocument();
        });

        it('file select allows file upload', async () => {
          const testData = [
            { x: 2, y: 3 },
            { x: 4, y: 12 },
          ];
          const str = JSON.stringify(testData);
          const blob = new Blob([str]);
          const file = new File([blob], 'test.json', {
            type: 'application/json',
          });

          const user = userEvent.setup({
            advanceTimers: jest.advanceTimersByTime,
          });
          render(<GameOfLife />, { wrapper: BrowserRouter });

          await act(async () => {
            await user.click(screen.getByTestId('open-import'));
          });

          user.upload(screen.getByTestId('import-select'), file);
          await waitFor(async () => {
            const importButton = screen.getByTestId('import-button');
            expect(importButton.getAttribute('disabled')).toBeNull();
          });

          expect(
            screen.getByTestId('import-select-label').textContent
          ).toContain('test.json');
        });

        it('sets the game configuration based on the uploaded file', async () => {
          const testData = [
            { x: 2, y: 3 },
            { x: 4, y: 12 },
          ];
          const str = JSON.stringify(testData);
          const blob = new Blob([str]);
          const file = new File([blob], 'test.json', {
            type: 'application/json',
          });
          File.prototype.text = jest.fn().mockResolvedValueOnce(str);

          const user = userEvent.setup({
            advanceTimers: jest.advanceTimersByTime,
          });
          render(<GameOfLife />, { wrapper: BrowserRouter });

          await act(async () => {
            await user.click(screen.getByTestId('open-import'));
          });

          user.upload(screen.getByTestId('import-select'), file);
          await waitFor(async () => {
            const importButton = screen.getByTestId('import-button');
            expect(importButton.getAttribute('disabled')).toBeNull();
          });

          await act(
            async () => await user.click(screen.getByTestId('import-button'))
          );

          let gameState = JSON.parse(
            screen.getByTestId('game-data').textContent!
          );

          expect(gameState).toEqual(testData);
        });
      });
    });
  });
});
