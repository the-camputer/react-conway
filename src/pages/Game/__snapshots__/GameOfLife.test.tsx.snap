// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOfLife renders game without crashing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-dxtbbu-JoyStack-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-display1 MuiTypography-colorNeutral MuiTypography-variantSolid css-16qvs2p-JoyTypography-root"
          id="title"
        >
          Conway's Game of Life
        </h1>
        <div
          class="sc-gsMHZj eiYyzm"
        >
          <canvas
            class="sc-bcHOTY kZGdfh"
            data-testid="gol-canvas"
            height="400"
            width="1880"
          />
        </div>
        <div
          data-testid="game-data"
        >
          [{"x":15,"y":2},{"x":16,"y":3},{"x":14,"y":4},{"x":15,"y":4},{"x":16,"y":4}]
        </div>
        <div
          class="MuiSheet-root MuiSheet-variantPlain MuiSheet-colorNeutral css-1bqfkf2-JoySheet-root"
        >
          <button
            class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeLg css-1mhx1qt-JoyButton-root"
            data-testid="play-game"
            type="button"
          >
            <span
              class="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="PlayArrowIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8 5v14l11-7z"
                />
              </svg>
            </span>
            Play
          </button>
          <button
            class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeLg css-1mhx1qt-JoyButton-root"
            data-testid="clear-game"
            type="button"
          >
            <span
              class="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ClearIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
            Clear
          </button>
        </div>
        <div
          class="MuiSheet-root MuiSheet-variantPlain MuiSheet-colorNeutral css-lmwrr5-JoySheet-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd css-1qd75b7-JoyFormControl-root"
          >
            <label
              class="MuiFormLabel-root css-4az1jf-JoyFormLabel-root"
              for=":r0:"
              id=":r0:-label"
            >
              Cell Size
               
              <p
                data-testid="cell-size"
              >
                40
              </p>
            </label>
            <span
              class="MuiSlider-root MuiSlider-variantSolid MuiSlider-colorPrimary MuiSlider-sizeLg css-1n12bqn-JoySlider-root"
            >
              <span
                class="MuiSlider-rail css-25lf7d-JoySlider-rail"
                data-testid="cell-size-rail"
              />
              <span
                class="MuiSlider-track css-1jq9w0k-JoySlider-track"
                style="left: 0%; width: 33.333333333333336%;"
              />
              <span
                class="MuiSlider-thumb css-1yv1cbk-JoySlider-thumb"
                data-index="0"
                style="left: 33.333333333333336%;"
              >
                <input
                  aria-orientation="horizontal"
                  aria-valuemax="100"
                  aria-valuemin="10"
                  aria-valuenow="40"
                  class="MuiSlider-input css-1aiebgt-JoySlider-input"
                  data-index="0"
                  max="100"
                  min="10"
                  step="2"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="40"
                />
              </span>
            </span>
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd css-1qd75b7-JoyFormControl-root"
          >
            <label
              class="MuiFormLabel-root css-4az1jf-JoyFormLabel-root"
              for=":r1:"
              id=":r1:-label"
            >
              Update Speed
               
              <p
                data-testid="update-speed"
              >
                500
              </p>
            </label>
            <span
              class="MuiSlider-root MuiSlider-variantSolid MuiSlider-colorPrimary MuiSlider-sizeLg css-1n12bqn-JoySlider-root"
            >
              <span
                class="MuiSlider-rail css-25lf7d-JoySlider-rail"
                data-testid="update-speed-rail"
              />
              <span
                class="MuiSlider-track css-1jq9w0k-JoySlider-track"
                style="left: 0%; width: 47.36842105263158%;"
              />
              <span
                class="MuiSlider-thumb css-1yv1cbk-JoySlider-thumb"
                data-index="0"
                style="left: 47.36842105263158%;"
              >
                <input
                  aria-orientation="horizontal"
                  aria-valuemax="1000"
                  aria-valuemin="50"
                  aria-valuenow="500"
                  class="MuiSlider-input css-1aiebgt-JoySlider-input"
                  data-index="0"
                  max="1000"
                  min="50"
                  step="10"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="500"
                />
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-dxtbbu-JoyStack-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-display1 MuiTypography-colorNeutral MuiTypography-variantSolid css-16qvs2p-JoyTypography-root"
        id="title"
      >
        Conway's Game of Life
      </h1>
      <div
        class="sc-gsMHZj eiYyzm"
      >
        <canvas
          class="sc-bcHOTY kZGdfh"
          data-testid="gol-canvas"
          height="400"
          width="1880"
        />
      </div>
      <div
        data-testid="game-data"
      >
        [{"x":15,"y":2},{"x":16,"y":3},{"x":14,"y":4},{"x":15,"y":4},{"x":16,"y":4}]
      </div>
      <div
        class="MuiSheet-root MuiSheet-variantPlain MuiSheet-colorNeutral css-1bqfkf2-JoySheet-root"
      >
        <button
          class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeLg css-1mhx1qt-JoyButton-root"
          data-testid="play-game"
          type="button"
        >
          <span
            class="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PlayArrowIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8 5v14l11-7z"
              />
            </svg>
          </span>
          Play
        </button>
        <button
          class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeLg css-1mhx1qt-JoyButton-root"
          data-testid="clear-game"
          type="button"
        >
          <span
            class="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ClearIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          Clear
        </button>
      </div>
      <div
        class="MuiSheet-root MuiSheet-variantPlain MuiSheet-colorNeutral css-lmwrr5-JoySheet-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd css-1qd75b7-JoyFormControl-root"
        >
          <label
            class="MuiFormLabel-root css-4az1jf-JoyFormLabel-root"
            for=":r0:"
            id=":r0:-label"
          >
            Cell Size
             
            <p
              data-testid="cell-size"
            >
              40
            </p>
          </label>
          <span
            class="MuiSlider-root MuiSlider-variantSolid MuiSlider-colorPrimary MuiSlider-sizeLg css-1n12bqn-JoySlider-root"
          >
            <span
              class="MuiSlider-rail css-25lf7d-JoySlider-rail"
              data-testid="cell-size-rail"
            />
            <span
              class="MuiSlider-track css-1jq9w0k-JoySlider-track"
              style="left: 0%; width: 33.333333333333336%;"
            />
            <span
              class="MuiSlider-thumb css-1yv1cbk-JoySlider-thumb"
              data-index="0"
              style="left: 33.333333333333336%;"
            >
              <input
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="10"
                aria-valuenow="40"
                class="MuiSlider-input css-1aiebgt-JoySlider-input"
                data-index="0"
                max="100"
                min="10"
                step="2"
                style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                type="range"
                value="40"
              />
            </span>
          </span>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd css-1qd75b7-JoyFormControl-root"
        >
          <label
            class="MuiFormLabel-root css-4az1jf-JoyFormLabel-root"
            for=":r1:"
            id=":r1:-label"
          >
            Update Speed
             
            <p
              data-testid="update-speed"
            >
              500
            </p>
          </label>
          <span
            class="MuiSlider-root MuiSlider-variantSolid MuiSlider-colorPrimary MuiSlider-sizeLg css-1n12bqn-JoySlider-root"
          >
            <span
              class="MuiSlider-rail css-25lf7d-JoySlider-rail"
              data-testid="update-speed-rail"
            />
            <span
              class="MuiSlider-track css-1jq9w0k-JoySlider-track"
              style="left: 0%; width: 47.36842105263158%;"
            />
            <span
              class="MuiSlider-thumb css-1yv1cbk-JoySlider-thumb"
              data-index="0"
              style="left: 47.36842105263158%;"
            >
              <input
                aria-orientation="horizontal"
                aria-valuemax="1000"
                aria-valuemin="50"
                aria-valuenow="500"
                class="MuiSlider-input css-1aiebgt-JoySlider-input"
                data-index="0"
                max="1000"
                min="50"
                step="10"
                style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                type="range"
                value="500"
              />
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
